envoy/certificate_provider/certificate_provider.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <list>
+
+#include "envoy/common/callback.h"
+#include "envoy/common/pure.h"
+#include "envoy/extensions/transport_sockets/tls/v3/cert.pb.h"
+#include "envoy/ssl/connection.h"
+
+#include "absl/strings/string_view.h"
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+namespace Envoy {
+namespace CertificateProvider {
+
+class OnDemandUpdateMetadata {
+public:
+  virtual ~OnDemandUpdateMetadata() = default;
+
+  virtual Envoy::Ssl::ConnectionInfoConstSharedPtr connectionInfo() const PURE;
+};
+
+using OnDemandUpdateMetadataPtr = std::unique_ptr<OnDemandUpdateMetadata>;
+
+class CertificateProvider {
+public:
+  struct Capabilites {
+    /* whether or not a provider supports generating identity certificates on demand */
+    bool provide_on_demand_identity_certs = false;
+  };
+
+  virtual ~CertificateProvider() = default;
+
+  virtual Capabilites capabilities() const PURE;
+
+  /**
+   * @return CA certificate used for validation
+   */
+  virtual const std::string trustedCA(const std::string& cert_name) const PURE;
+
+  /**
+   * Certificate provider instance which used to get tls certificates
+   * should provide at least one tls certificate.
+   * @return Identity certificates used for handshake
+   */
+  virtual std::vector<const envoy::extensions::transport_sockets::tls::v3::TlsCertificate*>
+  tlsCertificates(const std::string& cert_name) const PURE;
+
+  /**
+   * Add on-demand callback into certificate provider, this function might be invoked from worker
+   * thread during runtime
+   *
+   * @param metadata is passed to provider for certs fetching/refreshing
+   * @return CallbackHandle the handle which can remove that update callback.
+   */
+  virtual Common::CallbackHandlePtr
+  addOnDemandUpdateCallback(const std::string& cert_name, OnDemandUpdateMetadataPtr metadata,
```

Comment:
Is this actually used in this PR? I was looking at how listener/cluster would delay until the callback fires.

Reply:
No, this is not used in this PR. We are implementing a local mimic cert provider (code not submitted yet), and it will work with [bumping filter](https://github.com/envoyproxy/envoy/issues/22581) to mimic certs for downstream.

Reply:
Ok, thanks. I think this particular callback is quite interesting, you can do on-demand SDS, for example. Once it's hooked into upstream/downstream, the failure mode should be more clear (what happens if too many requests are made for certs, or one request times out).